### DISCUSSION TITLE
Inline small CSS in HTML (Parcel v2)

### DISCUSSION THREAD

**Discussion by hsablonniere** [parcel-bundler/parcel#5263] [Q&A]
Hello folks,

I'm using Parcel v2 with a static site generator.
All my articles reference the same stylesheet.
Some articles have custom CSS.
For "reasons", I author those as separate stylesheets referenced with `<link rel="stylesheet">`.

My usage of Parcel is really as an optimization phase so I would love if it could take an HTML as an entrypoint like this:

```html
<head>
<!-- ... -->
<link rel="stylesheet" href="common-big.css">
<link rel="stylesheet" href="specific-small.css">
</head>
```

And output this:

```html
<head>
<!-- ... -->
<link rel="stylesheet" href="common-big.css">
<style>/* contents of specific-small.css */</style>
</head>
```

Notes:

* Of course, deciding what's small enough to be inlined is open to debate but just having the possibility to do it in a custom plugin would be awesome!!
* I looked at these discussions: [1](https://github.com/parcel-bundler/parcel/discussions/5151) [2](https://github.com/Tom-Bonnike/netlify-plugin-inline-critical-css) [3](https://github.com/parcel-bundler/parcel/discussions/4765) [4](https://github.com/parcel-bundler/parcel/discussions/4708) without much success.
* I tried to use `@parcel/transformer-inline` and `@parcel/transformer-inline-string` but I end up having the inlined CSS in the `href` attribute.
* I looked at the source of [InlineStringTransformer](https://github.com/parcel-bundler/parcel/blob/v2/packages/transformers/inline-string/src/InlineStringTransformer.js) and [InlineTransformer.js](https://github.com/parcel-bundler/parcel/blob/v2/packages/transformers/inline/src/InlineTransformer.js). I see how I could create my "let's inline small CSS" but...
* I looked at the source of [HTMLPackager](https://github.com/parcel-bundler/parcel/blob/v2/packages/packagers/html/src/HTMLPackager.js) and I'm not sure if it know how to replace a `<link>` into a `<style>`.

Questions:

1. Do I make sense?
2. Do you think it could be useful to others?
3. Any idea what I could do? (custom plugin... ?)

Cheers :wink:

**Comment by hsablonniere** [comment]
I stumbled upon this https://github.com/xingqiao/parcel-plugin-inline-source but it's v1 and I don't want images and other assets...

**Comment by hsablonniere** [comment]
Same with this one: https://github.com/shff/parcel-plugin-inliner

**Comment by mischnic** [comment]
```html
<style>
@import "./style.css";
</style>
```

https://v2.parceljs.org/languages/html/#inline-script-and-style-tags

**Comment by hsablonniere** [comment]
@mischnic Thanks :+1: I may try that in the meanwhile but I also really want to find a way for Parcel to take a `<link rel="stylesheet" href="">` and output a `<style>`.

**Comment by hsablonniere** [comment]
Conclusion from trying to do it completely as a PostHTML plugin:

When using `@parcel/transformer-posthtml`, you cannot figure out the source path of the HTML you would need to resolve the CSS path an inject it.

I could create a PostCSS plugin that transforms `<link>` into styles as you described and let let Parcel do the rest but I would not be able to trigger than on asset size.

**Comment by hsablonniere** [comment]
Seems like patching the existing `@parcel/packager-html` is not that easy and maybe patching the `@parcel/transformer-html` or transforming before could work.

**Comment by hsablonniere** [comment]
@mischnic I ended up doing this: https://github.com/hsablonniere/hsablonniere.com/blob/master/local_modules/parcel-transformer-html-inline-css/parcel-transformer-html-inline-css.js

**Comment by devongovett** [comment]
You could maybe write a posthtml plugin to compile the link tag to a `<style>` tag like @mischnic showed about?

**Reply by hsablonniere** [reply]
Yed that's what I tried at some point. It works well but in a PostHTML plugin, I cannot figure out if the linked CSS is "big" or not to conditionally trigger inlining or not.

I think it's because my CSS URLs are relative to the HTML files. Therefore I have no proper way to use fs.stats to figure out the size of the CSS file.

I was wondering if and how the posthtml official plugin could pass the HTML source path. Maybe using the messages. If I had that, a simple postHTML plugin would work.
